ksql-parser/src/main/antlr4/io/confluent/ksql/parser/SqlBase.g4

```diff
@@ -144,6 +145,10 @@ groupingExpressions
     | expression
     ;
 
+values
+    : '(' (primaryExpression (',' primaryExpression)*)? ')'
+    ;
```

Comment:
I messed around with this and it's not trivial to breakup primaryExpression. rules cannot be re-used across expressions and expressions cannot be made of inline rules and other rules. e.g. both of these are invalid:
```antlr
foo : 
    a       #arule 
    | b     #brule
    ;

bar : 
    a       #arule 
    ;
```

```antlr
foo : 
    bar      
    | b     #brule
    ;

bar : 
    a       #arule
    ;
```

Reply:
I also can't split it up like:
```antlr

primaryExpression
    : literal
    | complexExpression
    ;

literal
    : NULL                                                                           #nullLiteral
    | number                                                                         #numericLiteral
    | booleanValue                                                                   #booleanLiteral
    | STRING                                                                         #stringLiteral
    ;

complexExpression
    : identifier STRING                                                              #typeConstructor
    | qualifiedName '(' ASTERISK ')'                              		               #functionCall
    | qualifiedName '(' (expression (',' expression)*)? ')' 						             #functionCall
    | CASE valueExpression whenClause+ (ELSE elseExpression=expression)? END         #simpleCase
    | CASE whenClause+ (ELSE elseExpression=expression)? END                         #searchedCase
    | CAST '(' expression AS type ')'                                                #cast
    | ARRAY '[' (expression (',' expression)*)? ']'                                  #arrayConstructor
    | value=primaryExpression '[' index=valueExpression ']'                          #subscript
    | identifier                                                                     #columnReference
    | identifier '.' identifier                                                      #columnReference
    | base=primaryExpression STRUCT_FIELD_REF fieldName=identifier                   #dereference
    | '(' expression ')'                                                             #parenthesizedExpression
    ;
```
because it doesn't let you have mutually left-recrusive rules (e.g. primaryExpression cannot show up in complexExpression)

Reply:
I added a test, though, to make sure that anything that isn't a literal fails

Reply:
I think you should be able to factor out the constant literals and use them in both `primaryExpression` and `insertIntoValues`. Something like this:

```antlr
primaryExpression
    : constantLiteral                                                                #constantLiteralExpr
    | identifier STRING                                                              #typeConstructor
    | qualifiedName '(' ASTERISK ')'                              		               #functionCall
    | qualifiedName '(' (expression (',' expression)*)? ')' 						             #functionCall
    | CASE valueExpression whenClause+ (ELSE elseExpression=expression)? END         #simpleCase
    | CASE whenClause+ (ELSE elseExpression=expression)? END                         #searchedCase
    | CAST '(' expression AS type ')'                                                #cast
    | ARRAY '[' (expression (',' expression)*)? ']'                                  #arrayConstructor
    | value=primaryExpression '[' index=valueExpression ']'                          #subscript
    | identifier                                                                     #columnReference
    | identifier '.' identifier                                                      #columnReference
    | base=primaryExpression STRUCT_FIELD_REF fieldName=identifier                   #dereference
    | '(' expression ')'                                                             #parenthesizedExpression
    ;

insertIntoValues
    : '(' constantLiteral (',' constantLiteral)* ')'
    ;

constantLiteral
    : NULL                                                                          #nullLiteral
    | number                                                                        #numericLiteral
    | booleanValue                                                                  #booleanLiteral
    | STRING                                                                        #stringLiteral
    ;

insertIntoColumns
    : '(' identifier (',' identifier)* ')'
    ;
```

Reply:
Nice!